### ISSUE TITLE
Add Block Placement

### ISSUE THREAD

**Issue by controllerface** [controllerface/bvge#78]
Now that I can break blocks, and collect the resulting resources, it is now time to actually implement block placement. I anticipate this might take a bit of work to get right as there's several moving parts.

I have already started in on some of the prerequisite things to make this work, namely an event system. I had to add this for the basic HUD implementation, to inform that renderer when the player's inventory changed, and I think I will use it for placement of blocks as well, in addition to using an ECS component.

I am not sure exactly of the order I need to implement each thing, but the rough outline is this:
- handle the events NEXT_ITEM and PREV_ITEM which will cycle through the blocks that are currently placeable
- handle some kind of CANCEL event that will exit "placement mode" 
- when in placement mode with a specific block, spawn a "ghost" version of the block
- the ghost block should be bound to the mouse cursor, so the player can move it around
- when cycling through blocks to place, the previous ghost block should be destroyed, and the new one should take its place
- when cancelling, the current block should just be destroyed
- when the player click the left mouse button while a ghost block is present, it must place the block
- when a block is placed, it must deduct the appropriate amount of the mineral required for it
- if placing a block results in insufficient inventory to place another, it should auto-cancel placement mode
  - otherwise, a new ghost block of the same type should be spawned

I will probably want to add more points as I go, but this is the general idea for now. I want to try and ensure I make this logic expandable to other items that i may add later, like various crafted objects, etc. so I will try and keep things tied to an "entity" rather than a block specifically, though I will make this more of a goal than a hard rule. There may be a need for multiple modes later, but as long as the code that allows the objecting being placed to follow the mouse is agnostic, that should be enough.

**Comment by controllerface** [comment]
I got a decent way into an implementation and even had blocks spawning in the game world, however I realized I have a problem, and I need to decide on the best solution.

I implemented the block placement by spawning a block and effectively pinning it to the mouse cursor. I reused the mouse entity update kernel, which adjusts the first point in the hull to match the mouse. This resulted in the block kind of "dangling" from it's corner, pinned to the mouse position. I actually was kind of happy with that as a starting point because the object still has physics and everything when pinned this way, so If I wanted to make use of that later, I could simply figure out how to _avoid_ it for block placing, and just re-enable it later.

However, I quickly realized my implementation was flawed. What I did to track the element was store the ID of the entity in the world sector memory. That unfortunately doesn't work because cycling through more than one object caused issues with ID offsets as the object would get deleted and replaced and the order mattered. I had just figured this out and was about to fix the ordering issue when I noticed another issue. Having a block out and attached to the mouse, then moving the screen until it unloads sectors, drops the block and pins some random block to the mouse instead.

So I realized the entity ID isn't going to work, because it changes at runtime. I have a few options though for handling it:

1. I can re-think using world spawned objects and instead in the HUD renderer, grab the mouse co-ordinates and render the current block as a simple graphic. Doing it this way loses the world/physics interactions, and would have to work very much like terraria. This might be the "easiest" option, however I can also imagine some issues there, mainly because I would need to work out how to do collision in this mode to ensure blocks can't be placed directly on top of others.
2. I can make a persistent "ghost block" that is always present at the beginning of the buffer and instead of spawning/despawning the block, I could simply mutate its properties to "be" whatever block I am trying to place. Then when the block is placed, then a copy could be placed at the position of the ghost block. because this block would be in a persistent position, it would not have the issues described above, and also would not require spawning/despawning objects, only spawning them when placed. 
3. I can add unique IDs, which I need to do for other reasons anyway, and instead of tracking the entity ID, use those unique IDs, which would not change due to world updates, entities would always contain the same unique ID value no matter their entity ID. This would require carefully deciding what data type the ID should be, as it would limit the number of entities that could be placed in the world.

**Comment by controllerface** [comment]
I went with the second option and it's working well so far. I haven't yet made it possible to make the block spawn in the world, but the "ghost" block is working, I can collect blocks and then cycle through all the collected block types using the mouse wheel + ctrl, and can also select "none" which is the default state. When there is no block selected, the ghost block is simply not visible. It does still technically exist, and I do need to work on disabling it in the physics sim when it is disabled for rendering, but the visual effect is good. 

I now need to do a few things: 
- the block and in fact probably the inner mouse cursor as well, should be restricted to the player's range
- I need a way to rotate the block as well as lock/snap to the initial rotation state (i.e. zero rotation)
- the left click action needs to have awareness of whether or not a block to place is present
  - if it is, instead of a punch the block should be spawned and resources deducted from inventory
- when a block is placed, a check should be made to ensure there is enough material for another block
  - if not, a deselect event should be emitted

**Comment by controllerface** [comment]
In my haste, I forgot to address some issues with the ghost block, specifically how it collides with other objects. Even when not visible, it was causing other hulls to collide with it, making the player entity jitter when jumping and causing broken block pieces to mysteriously float around. I have addressed this by checking for the ghost block in the kernels and restricting collision reactions ONLY tot he ghost block itself. I want this to be the case so that placing the block in the world will still be bound by physics, so when placement actually works it won't be possible to place it inside other objects.

One thing I need to look at is how the ghost block tracks the mouse. Right now, it simply follows it directly, or more precisely, the first point in the block is "pinned" to the mouse. It actually should _move toward_ the mouse instead, which would cause it to collide with other blocks instead of just erratically clipping though them as it does now. I will probably leave this until after I actually have placement and other things in place, but it will be needed before this is fully "done".

**Comment by controllerface** [comment]
Still not done on this yet, but have more thoughts I want to get down:

As I am going through this process, I am finding there's some things unrelated to this work directly that I may want to reconsider before moving on because it will have an effect on how this and other things get implemented. Mainly, I have gone back and forth a bit on whether the ECS class is really worth having, but now I think I see the value in it. I need to detect if the player is currently in "block placing mode" from within the physics system, but would it really be a good idea to pass in the player inventory for this? 

No, I don't think so. It would be better to decouple those things and have that as a component. Now, the place where I need to do that is inside the entity control method, which also is where the state data is updated. Taking a second look here, I now am wondering if it was really wise to do all this work GPU side instead of CPU side. I am already paying a cost to transfer data into the GPU buffer, why not just do all the logic for the state changes CPU side, where the enum and util classes make it more straight forward?

Honestly, I think I need to re-consider how much logic I offload into the GPU. I still do want to keep any logic that needs to run in parallel over there as it really has a huge performance benefit. But for processes that only operate on the player, and which must incur a GPU penalty anyway, it's probably better to do the branching logic on the CPU where it can be made more readable.

I have added a few todo's to address this, and also some places where I am currently looping the ECS data when I really just need to grab the player data. I don't think this is causing any huge performance problems, but it certainly is more convoluted than it needs to be. I want to make sure I'm using the best tool for the job and not just cramming stuff into the GPU when it's not necessary. As I consider that I may want to migrate from OpenGL/CL to Vulkan some day, any code that is CPU side is one les thing that needs porting.

**Comment by controllerface** [comment]
I now have the ability to place blocks in the world. It's still not completely right, but the basic spawning process works. The placed block is spawned directly on the mouse cursor location instead of the block location, and does not have any rotation applied. I will need to write a simple kernel that just copies the block cursor data into the newly spawned blocks' buffers. I also need to make sure that the proper amount of resources is removed from the player inventory, as well as add checks to ensure that once the player doesn't have enough resources, the block is deselected. 

In all honesty, getting this working was about 5% of the time I spent on this tonight, I spent most of my time looking over some existing stuff like the ECS and related classes, and also renaming some things to try and make the intent a bit clearer. I am getting a bit sidetracked lately as I interact with the codebase a bit more, which I think is OK because I want to get a feel for what works and what doesn't. As I go, I find that (unsurprisingly) sticking to a "pure" ECS system is not going to work for me.

Most discussion I find around ECS involves a lot of "theory" about how such a system should work. Aside from it being almost exclusively focused on C/C++ specific characteristics, I find very little good discussion on practical uses of general ECS designs. There's quite a bit about engine specific implementations of it (i.e. Unity/Unreal) and even there, it seems like the implementations are not completely "pure" either, in the sense that there's usually some kind of exception for a special case or two. I also find that since I have a lot of data on GPU, and relatively little CPU side, my use case doesn't quite fit the mold. 

The benefits of ECS as stated online boil down to two main aspects; cache locality and decoupling. The former simply doesn't apply in Java, at least with normal classes. Once could conceivably use FFM to force structures into certain memory alignments, but I have negative interest in doing this. When Valhalla comes out, that may change things here but for now I'm not really thinking about this. 

The decoupling aspect though, is proving to be pretty helpful, though perhaps a little knee-capped for me since I'm actually only using the ECS class for the player right now, all other entities are handled 100% GPU-side. Still, I like not having to make all player related data accessible to every class that uses it, allowing them to query the ECS for it is much better. Though at the same time, there's a couple things that still need to be passed around, like the player's inventory.

I am definitely not using a pure design either, I hit the limitations of the "only data containers" restriction on components almost as soon as I added them. Again, this is very much a restriction that I think applies more to scenarios where the ECS is used in non-Java languages. Indeed, I would guess that the majority of people who developed the ECS paradigm would dismiss using a JVM language out of hand, and all the baggage that goes along with that attitude. I think it is worth working with the strengths of the language I am in then, and not adhering to any specific dogma about what an ECS is or isn't. In that vein, I may just stop calling it ECS and just do whatever I want 😆

**Comment by controllerface** [comment]
Block placing is a little more sane now. When placing the block, the newly placed block adopts the exact position/rotation data of the block cursor. The proper amount of resources is also deducted from the player inventory and the HUD is updated to properly reflect the new state after placing. Also, if the player's resource amount falls under the required level (which is 4) block selection mode is exited directly. 

I had to do a few things to make this process a little more user friendly as well, such as preventing the punch action from occurring when placing a block as well as preventing it when the resource count falls below the minimum. In both cases, the player would easily start destroying the newly placed block, which is not ideal. This entailed adding "latch" and "latch required" flags to the input class.

I still need to add some way to lock the rotation of the block as well as snap it to the sector/block grid. I do want to make it possible to adjust these values if desired, but I think it probably makes more sense to have both of those be done via some kind of mode switch, and default to a locked/snapped placement. I would also like to make it possible to spawn a block as non-static, so it will drop to the ground. Probably not super useful, but could be fun to mess around with.

I do also need to figure out a way to restrict the block placement to the player's range, similarly to the punch/destroy range. Along with that, I want to change the block cursor logic so it seeks to the mouse instead of being set directly to it, so it won't end up inside other geometry. I may need to get creative to achieve this though, as it would require something like the player update to loop to implement. I am not quite ready for non-player entity loops yet so I may try to work out a different solution that achieves the same result. At the end of the day, the thing I want to avoid is placing the block inside another block.

**Comment by controllerface** [comment]
i have implemented a simple but effective block snapping function and applied it to the block position when updating. This already works a lot better and makes it very easy to place blocks against existing ones and have them line up correctly. I will note that due to the very minor screen movement introduced by the player's idle animation, there can be extremely tiny gaps between blocks. It kind of has the same effect that all the generated blocks do, in that sometimes it looks as though the edges of blocks flash quickly. For now, I'm OK with this though, as I will refine all of these systems later, and I just want to get the basic idea working.

I don't currently have any direct rotation locking in place, but moving blocks directly next to others has the effect of automatically doing this, so I'm in less of hurry to add this now, and will probably close this and move on to the next thing. I am still considering this a prototype so rather than obsess about every individual feature, I want to keep pushing forward and refine later.
